### PR TITLE
Feature Implement QLearner

### PR DESCRIPTION
**Pull request by usaito**
## New Features

- implement QLearner in the policy module, which first estimates the expected reward function (or the q function) from the logged bandit data, and then use that estimator to make new decisions on the test data (such as via argmax or softmax)
- See example 1 of in Appendix A of https://arxiv.org/pdf/2002.08536v2.pdf for the definition of QLearner implemented here
- add some tests corresponds to the inputs and performance of the QLearner class in `tests/policy`
- implement "Gumbel softmax trick" (see Section 4.1 of https://arxiv.org/pdf/2105.00855.pdf) in IPWLearner/QLearner/NNPolicyLearner (of `obp.policy`) to efficiently sample a ranking of actions from the plucket luce ranking distribution

## Minor Fix
- unify some test functions to delete repetitive training of a same policy in `tests/policy/test_offline_learner_performance.py` to reduce the time needed to run the tests
- fix some numpy/pandas/sklearn warnings (now the tests will run without any warnings)

### CODE REVIEW

File: obp/dataset/synthetic_slate.py [hunk 1]
```diff
@@ -1433,7 +1433,7 @@ def linear_behavior_policy_logit(
         Controls the random seed in sampling dataset.
 
     tau: int or float, default=1.0
-        A temperature parameter, controlling the randomness of the action choice.
+        A temperature parameter, controlling the randomness of the action choice by scaling thescores before applying softmax.
```
Comment (fullflu) [inline obp/dataset/synthetic_slate.py:1436]: [nits]
`thescores` -> `the scores` ?

File: obp/utils.py [hunk 2]
```diff
@@ -794,7 +794,7 @@ def check_rips_inputs(
 
 def sigmoid(x: Union[float, np.ndarray]) -> Union[float, np.ndarray]:
     """Calculate sigmoid function."""
-    return 1.0 / (1.0 + np.exp(-x))
+    return np.exp(np.minimum(x, 0)) / 1.0 / (1.0 + np.exp(-np.abs(x)))
```
Comment (fullflu) [inline obp/utils.py:797]: [nits]

`np.exp(np.minimum(x, 0)) / (1.0 + np.exp(-np.abs(x)))` seems better (typo?)

**Review by fullflu — COMMENTED** [review]
nits (no warning happened)

I'll add a review of QLearner in the next comment.

**Comment by usaito** [comment]
@fullflu Thanks!

**Review by fullflu — COMMENTED** [review]
Review of QLearner

File: obp/policy/offline.py [hunk 3]
```diff
@@ -281,24 +283,21 @@ def sample_action(
         Returns
         -----------
         action: array-like, shape (n_rounds_of_new_data, n_actions, len_list)
-            Action sampled by a trained classifier.
+            Ranking of actions sampled by the Gumbel softmax trick.
```
[the remark this answers is not in the document]
Comment (usaito) [reply obp/policy/offline.py:286]: @fullflu Thanks! I'll replace the variable name from `sampled_action` to `sampled_ranking`

**Comment by usaito** [comment]
@fullflu Thanks again for your thoughtful comments! I'll address some recent minor comments in the following ongoing PR.
https://github.com/st-tech/zr-obp/pull/145
